hera_filters/dspec.py

```diff
@@ -1661,43 +1680,102 @@ def _fit_basis_1d(x, y, w, filter_centers, filter_half_widths,
     info['basis_options'] = basis_options
     info['amat'] = amat
     info['skipped'] = False
-    if method == 'leastsq':
-        a = np.atleast_2d(w).T * amat
-        try:
-            res = lsq_linear(a, w * y)
-            cn_out = res.x
-        # np.linalg.LinAlgError catches "SVD did not converge."
-        # which can happen if the solution is under-constrained.
-        # also handle nans and infs in the data here too.
-        except (np.linalg.LinAlgError, ValueError, TypeError) as err:
-            warn(f"{err} -- recording skipped integration in info and setting to zero.")
-            cn_out = 0.0
-            info['skipped'] = True
-    elif method == 'matrix':
+    if _are_wgts_binary(w):
+        # Build mask and flags arrays
+        mask = np.array(w, dtype=bool)
+        flags = np.logical_not(mask)
+
+        # Pre-compute expected XTX matrix without flags for faster computation later
+        square_key = _fourier_filter_hash(filter_centers=filter_centers, filter_half_widths=filter_half_widths,
+                                    x=x, hash_decimal=hash_decimal, label='covariance')
         fm_key = _fourier_filter_hash(filter_centers=filter_centers, filter_half_widths=filter_half_widths,
-                                      filter_factors=suppression_vector, x=x, w=w, hash_decimal=hash_decimal,
-                                      label='fitting matrix', basis=basis)
-        if basis.lower() == 'dft':
-            fm_key = fm_key + (basis_options['fundamental_period'], )
-        elif basis.lower() == 'dpss':
-            fm_key = fm_key + tuple(nterms)
-        fmat = fit_solution_matrix(w, amat, cache=cache, fit_mat_key=fm_key)
-        info['fitting_matrix'] = fmat
-        cn_out = fmat @ y
+                                    filter_factors=suppression_vector, x=x, w=w, hash_decimal=hash_decimal,
+                                    label='fitting matrix', basis=basis, mode=method)
+        if square_key in cache:
+            covmat = cache[square_key]
+        else:
+            covmat = np.dot(np.conj(amat).T, amat)
+            cache[square_key] = covmat
 
-    elif method == 'solve':
-        try:
-            A = (np.conj(amat).T * w) @ amat
-            b = np.conj(amat).T @ (w * y)
-            cn_out = np.linalg.solve(A, b)
+        if not fm_key in cache:
+            XTX = covmat - np.conj(amat[flags]).T @ amat[flags]
+
+        Xy = np.conj(amat[mask]).T @ y[mask]
 
-        except (np.linalg.LinAlgError, ValueError, TypeError) as err:
-            warn(f"{err} -- recording skipped integration in info and setting to zero.")
-            cn_out = 0.0
-            info['skipped'] = True
+        if method == 'matrix':
+            if fm_key in cache:
+                XTXinv = cache[fm_key]
+            else:
+                XTXinv = np.linalg.pinv(XTX)
+                cache[fm_key] = XTXinv
 
+            cn_out = np.dot(XTXinv, Xy)
+
+        elif method == 'solve':
```

Comment:
Is it fair to say that the 'solve' method is only supported for binary weights? If so, maybe we should add a valueerror if non-binary weights are provided for solve.

Reply:
It should also work on non-binary weights. On line 1763, I have a version for non-binary weights.

Reply:
I did just realize the different modes are not in the same order for binary and non-binary weights. If you’d like me to make the order consistent for binary and non-binary weights, I’m happy to do that

Reply:
I see this now. I think it is very clear that this is supported when viewing the file in it's entirety rather than the diff mode where it isn't clear. I think the confusion is an artifact of githubs visual layout so no need to modify the code further.